### ISSUE TITLE
"HTTP 409 Conflict" when requesting /transmission instead of /transmission/web

### ISSUE THREAD

**Issue by gdubicki** [transmission/transmission#231]
I am using Transmission 2.92 (14714) (but I am pretty sure I have seen this on maaany versions before).

Sometimes I enter too short URL in my browser and land on the unfriendly HTTP 409 page instead of Transmission web UI.

I have reproduced it with curl:

```
root@host:~# curl -H "Authorization: Basic (...)" -v http://192.168.100.100:9091/transmission/
> GET /transmission/ HTTP/1.1
> Host: 192.168.100.100:9091
> User-Agent: curl/7.51.0
> Accept: */*
> Authorization: Basic (...)
>
< HTTP/1.1 409 Conflict
< Server: Transmission
< X-Transmission-Session-Id: dfqzFD8fdAiuw11jBHj37FSb5kuQfcYQ5DfjuQit89aAa9Jh
< Date: Mon, 27 Mar 2017 22:16:52 GMT
< Content-Length: 580
< Content-Type: text/html; charset=ISO-8859-1
<
<h1>409: Conflict</h1><p>Your request had an invalid session-id header.</p><p>To fix this, follow these steps:<ol><li> When reading a response, get its X-Transmission-Session-Id header and remember it<li> Add the updated header to your outgoing requests<li> When you get this 409 error message, resend your request with the updated header</ol></p><p>This requirement has been added to help prevent <a href="http://en.wikipedia.org/wiki/Cross-site_request_forgery">CSRF</a> attacks.</p><p><code>X-Transmission-Session-Id: dfqzFD8fdAiuw11jBHj37FSb5kuQfcYQ5DfjuQit89aAa9Jh</code></p>
```
It's not really a big deal but it would be nice if this could be polished.

**Comment by killemov** [comment]
Works as designed. Transmission expects a token, or X-Transmission-Session-Id, in the header. Repeat the request with the given token and you're golden.

**Comment by gdubicki** [comment]
Maybe Transmission should rather serve it's API under different path (`/api` ?) than it web UI (`/web`)? That would solve it too.

**Comment by fernandog** [comment]
i get this when Transmission GUI is updating UI and my app is sending a RPC command

**Comment by jquince** [comment]
Apparently the trail / is important
http://x.x.x.x:9091/transmission/web 409
this works -> http://x.x.x.x:9091/transmission/web/

**Comment by kopax** [comment]
I finally managed to access the webui by adding a `/` after `/transmission/web`, but the `/transmission/rpc` is still returning 404.

The transmission udp/tcp port is not visible publicly, is that an issue for the web client? How do I export UDP/TCP? How do I get the RPC link properly?

**Comment by kopax** [comment]
@jquince this help but then the ajax POST request to `/transmission/rpc` is failing with 404 in my case.

I only opened 443 from traefik https reverse proxy publicly, any tought?

**Comment by kopax** [comment]
Guys, if you are using Traefik v2, it is important to set 

```diff
+ PathPrefix(`/transmission`)
# instead of
- PathPrefix(`/transmission/web`)
```

Otherwise the RPC POST request will fail with 404.

**Comment by toxic0berliner** [comment]
> ```diff
> PathPrefix(`/transmission`)
> ```

Hmm, where do you set this PathPrefix ? I'm starting transmission inside docker and the service and rotuer get auto-created, I could add labels to the docker container to have traefik add things to it, but where should I add this PathPrefix ? Router ? Service ? A bit more of this file would help me to find out ;)

Thanks in advance @kopax

**Comment by kopax** [comment]
@toxic0berliner, it's a traffic settting that goes on the routeur, for example:

```json
{
  "traefik.http.routers.transmissionGateway.rule": "Host(`get.kopaxgroup.com`) && PathPrefix(`/transmission`)"
}
```

This is my whole traefik v2 configuration:


```json
{
            "traefik.enable": "true",
            "traefik.passHostHeader": "true",
            "traefik.http.routers.transmissionGateway.entrypoints": "web-secure",
            "traefik.http.routers.transmissionGateway.tls.certresolver": "gandiLetsencrypt",
            "traefik.http.routers.transmissionGateway.tls.domains[0].main": "get.mydomain.com",
            "traefik.http.routers.transmissionGateway.rule": "Host(`get.mydomain.com`) && PathPrefix(`/transmission`)",
            "traefik.http.middlewares.redirect-to-https.redirectscheme.scheme": "https",
            "traefik.http.routers.transmissionGatewayRedirs.rule": "hostregexp(`{host:.+}`)",
            "traefik.http.routers.transmissionGatewayRedirs.entrypoints": "web",
            "traefik.http.routers.transmissionGatewayRedirs.middlewares": "redirect-to-https"
}
```

I am happy to help. If you can find **how to redirecto from `/` to `/transmission`, that would help me have a 200 status code on the root path, instead of the current 404.**

**Comment by toxic0berliner** [comment]
Thanks @kopax 
Sadly, this does not solve my issue, since my router gets defined properly by the docker provider already, not using PathPrefix.

I also have redirect from / to /transmission working by default since transmission is doing it itself.
I suspect it is exactly your PathPrefix rule on the router that prevents / from beeing matched to this router, so just try removing it and make / match to your transmission router, it will forward to the transmission service on / that will issue a HTTP 301 to redirect you to /transmission/web/

On my side, My current issue is that I get a http 409 Conflict on each of the rpc POST, I suspect it's some of my middleware or some header that traefik is adding since bypassing traefik does not produce this error...
 Will look into it when I get some time.

**Comment by kopax** [comment]
@toxic0berliner thanks for your feedback and for the redirection information I didn't have.

I fixed the RPC 409 error WITH the `PathPrefix`. If you can get to there just for testing. If you find how to get both the redirect and the `PathPrefix` working together, please let me know. You can contact me on freenode if you need further assistance on my setup `/msg dka`.

**Comment by toxic0berliner** [comment]
Will try asap @kopax .
I think just adding another PathPrefix match on / alone should do the trick for you.

**Comment by toxic0berliner** [comment]
Hello again.
I tried adding the PathPrefix thing, it did not change anything for me.
Looking at the dev-tool in Firefox, I believe transmission is setting a custom http header X-transmission-sessionID or similar.
When browsing transmission directly without traefik, I see a first 409 conflict on the rpc url, then all following reuests to rpc url have this transmission session id http header.
I get that it probably should be a cookie instead of a custom header, but nonetheless, I have found no way to tell traefik to let this http header through...

Any way to avoid using such a custom http header in transmission ? Had to switch to qBittorrent for the time beeing.

**Comment by erosenberg** [comment]
Here was my workaround:

```
http:
  routers:
    transmission:
      entryPoints:
        - https
      rule: Host(`transmission.<custom domain>`)
      service: transmission-svc
      tls: {}
  services:
    transmission-svc:
      loadBalancer:
        servers:
          - url: "http://<host ip>:9091/"
          
```

**Comment by kopax** [comment]
Hi @toxic0berliner, it's weird, I use transmission for a while with my configuration, have you tested it as it is?

**Comment by toxic0berliner** [comment]
Yes @kopax I just re-checked since I had a typo, but I have this as labels : 
````
traefik.http.routers.transmission | Host(`transmission.mydomain.com`,`transmission.myOtherdomain.com`) && PathPrefix(`/transmission`,`/`)
traefik.http.services.transmission.loadBalancer.server.port | 9091
````
And I still get a 409 Conflict message on the RPC call to https://transmission.mydomain.com/transmission/rpc

**Comment by kopax** [comment]
Try with:

```bash
traefik.http.routers.transmission | Host(`transmission.mydomain.com`) && PathPrefix(`/transmission`)
```

There's no reason that justify why it work for me and not for you.

**Comment by toxic0berliner** [comment]
I found the source of the issue. It seems it is "normal behavior" to get a 409 but I believe transmission UI is actually fetching some information from the 409 response, the transmission sessionID in fact.
And I had a error-pages middleware on traefik that ould replace that 409 error page by another one, removing this sessionID and therefore it did not work.
I would realy advise not to rely on http error codes to send data between UI & back. The issue would not arise if GUI got the sessionID from a requests that results in a code 200. I might be mistaken but relying on data inside a 409 response does not seem best practice to me.

**Comment by teras** [comment]
I am getting a similar issue when I put transmission behind a nginx proxy and try to access it. The code I am using for proxy is

```
    location / {
        proxy_pass http://192.168.1.148:9091/transmission/web/;
        proxy_set_header    X-Real-IP  $remote_addr;
        proxy_set_header    X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header    Host $http_host;
        proxy_redirect false;
    }
```

If I use as `proxy_pass`  the value of  `http://192.168.1.148:9091` instead, and access the server as `https://myserver/transmission/web/` everything is fine.

If I use the above `proxy_pass`, and try to use the server as  `https://myserver/`, then I also get a `409` error.

Is there a way to access the server directly through the name, and not having to write the path?

**Comment by GaryElshaw** [comment]
Does this help? https://github.com/transmission/transmission/issues/899#issuecomment-1462030317

**Comment by teras** [comment]
> Does this help? [#899 (comment)](https://github.com/transmission/transmission/issues/899#issuecomment-1462030317)

I feel so stupid that didn't find it earlier.

Indeed this worked. The find tuning required, for future reference is to use this:

```
     location = / {
        return 301 /web/;  # The slash at the end is important.
    }

    location / {
        proxy_pass http://192.168.1.148:9091/transmission/;
```
 followed but the rest of the configuration. Now I can use it as desired; both with `https://myserver/web/` as well as `https://myserver/`

THANK YOU
